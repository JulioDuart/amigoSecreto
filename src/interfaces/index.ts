export interface IAmigos {
    nome: string,
    id: number,
    sorteado: boolean,
    escolheu: boolean,
    tirou: string
}