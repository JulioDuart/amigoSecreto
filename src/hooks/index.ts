import { useRecoilState, useSetRecoilState } from "recoil"
import { IAmigos } from "../interfaces"
import { statetListaAmigos, statetListaAmigosSorteados } from "../state"
import shuffle from "just-shuffle"
import { realizarSorteio } from "../helpers/realizarSorteio"

export const useListaAmigos = () => {
    const amigos = useRecoilState<IAmigos[]>(statetListaAmigos)

    return amigos[0]
    
}

export const useSetListaAmigos = () => {
    const setAmigos = useSetRecoilState<IAmigos[]>(statetListaAmigos)
    const amigos = useRecoilState<IAmigos[]>(statetListaAmigos)

    return (amigo:IAmigos) => {
        const jaTem = amigos[0].find(a => a.nome === amigo.nome)
        if (jaTem) throw Error("Nome jÃ¡ cadastrado!")
        else setAmigos(amigosAntigos => [...amigosAntigos, amigo])
    }
}

export const useSorteioAmigo = () => {
    const listaAmigos = useListaAmigos()

    return (id:number|undefined) => {
        const amigosNaoSorteados = listaAmigos.filter(a => a.sorteado === false)
        const amigosFiltrado = amigosNaoSorteados.filter(a => a.id !== id)
        const numeroRestante = amigosFiltrado.length
        return amigosFiltrado[Math.floor(Math.random() * numeroRestante)]
    }
}

export const useListaAmigosQueEscolheram = () => {
    const amigos = useRecoilState<IAmigos[]>(statetListaAmigos)

    const amigosNaoSorteados = amigos[0].filter(a => a.escolheu === false)
    return amigosNaoSorteados 
}

export const useSortearAmigo = () => {
    const listaAmigos = useListaAmigos()
    const listaAmigosModificada:IAmigos[] = shuffle(listaAmigos)

    const restult = useSetRecoilState(statetListaAmigosSorteados)

    return () => {
        const resultado = realizarSorteio(listaAmigosModificada)
        restult(resultado)
    }
}

export const useListaAmigosQueSorteados = () => {
    const amigos = useRecoilState<IAmigos[]>(statetListaAmigos)

    return () => {
        const amigosNaoSorteados = amigos[0].filter(a => a.sorteado === false)
        return amigosNaoSorteados
    }
}

export const useListaAmigosSorteados = () => { 
    const amigos = useRecoilState(statetListaAmigosSorteados)
    return amigos[0]   
}

/*
export const useSortearAmigo = () => {
    const listaAmigos = useListaAmigos()


    const setAmigos = useSetRecoilState<IAmigos[]>(statetListaAmigos)

    return (idSorteado:number|undefined, idEscolheu:number|undefined) => {
        const amigoFiltrado = listaAmigos.find(a => a.id === idSorteado)
        if (amigoFiltrado) {
            const nome = amigoFiltrado.nome
            const atuAmigoSort = {
                id: amigoFiltrado.id,
                nome: nome,
                sorteado: true,
                escolheu: amigoFiltrado.escolheu,
                tirou: amigoFiltrado.tirou
            }
            const index = listaAmigos.findIndex(a => a.id === idSorteado)
            setAmigos(amigosAntigo => [...amigosAntigo.slice(0, index), atuAmigoSort, ...amigosAntigo.slice(index + 1)])
            const amigoFiltradoEscolheu = listaAmigos.find(a => a.id === idEscolheu)
            if (amigoFiltradoEscolheu) {
                const nome = amigoFiltradoEscolheu.nome
                const atuAmigoEsco = {
                    id: amigoFiltradoEscolheu.id,
                    nome: nome,
                    sorteado: amigoFiltradoEscolheu.sorteado,
                    escolheu: true,
                    tirou: amigoFiltrado.nome
                }
                const index = listaAmigos.findIndex(a => a.id === idEscolheu)
                setAmigos(amigosAntigo => [...amigosAntigo.slice(0, index), atuAmigoEsco, ...amigosAntigo.slice(index + 1)])
            }
        }
        
    }
}

*/