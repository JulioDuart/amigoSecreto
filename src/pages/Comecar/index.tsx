import { useState } from "react"
import style from "../componentes.module.scss"
import { IAmigos } from "../../interfaces"
import { useListaAmigos } from "../../hooks"
import { useNavigate } from "react-router-dom"
import Formulario from "./componentes/FormularioInput"

const sacola = require("../../assets/sacolas.png")
const playSVG = require("../../assets/play.png")

export default function Comecar() {

    const listaAmigos:IAmigos[] = useListaAmigos()

    const nav = useNavigate()

    const botao = () => {
        if (listaAmigos.length > 3) {
            nav("/sortear")
        } 
    }

    return (
        <div className={style.divPrincipal} >
            <h1>Vamos come√ßar!</h1>
            <div className={style.divInput} >
                <Formulario />
            </div>
            <div className={style.divLista} >
                {listaAmigos?.map(a => {
                    return (
                        <p key={a.id} role="listaItens" >{a.nome}</p>
                    )
                })}
            </div>
            <div className={style.divFooter}>
                <button onClick={botao} role="btn_enviar" disabled={listaAmigos?.length > 3 ? false : true} >
                    <img src={playSVG} alt="play" style={{ height: 20, width: 20, marginRight: 20 }} />
                    <p>Iniciar brincadeira!</p>
                </button>
                <img src={sacola} alt="sacola" />
            </div>
        </div>
    )
}