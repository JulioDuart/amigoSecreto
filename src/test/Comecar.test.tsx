import React from "react";
import { RecoilRoot } from "recoil";
import Comecar from "../pages/Comecar";
import { fireEvent, render, screen } from "@testing-library/react";
import { useListaAmigos, useSetListaAmigos } from "../hooks";
import { IAmigos } from "../interfaces";

const mockNavegacao = jest.fn()

jest.mock("react-router-dom", () => {
    return {
        useNavigate: () => mockNavegacao
    }
})

describe("descreve a pagina inicial", () => {

    test("deve ser renderizada corretamente", () => {
        const {container} = render(
            <RecoilRoot>
                <Comecar />
            </RecoilRoot>
        )

        expect(container).toMatchSnapshot()
    })
})