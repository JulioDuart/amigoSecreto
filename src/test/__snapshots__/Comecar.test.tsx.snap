// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`descreve a pagina inicial deve ser renderizada corretamente 1`] = `
<div>
  <div
    class="divPrincipal"
  >
    <h1>
      Vamos come√ßar!
    </h1>
    <div
      class="divInput"
    >
      <form>
        <input
          placeholder="Adicionar nome dos participantes"
          value=""
        />
        <button
          disabled=""
          type="submit"
        >
          Adicionar
        </button>
      </form>
    </div>
    <div
      class="divLista"
    />
    <div
      class="divFooter"
    >
      <button
        disabled=""
        role="btn_enviar"
      >
        <img
          alt="play"
          src="play.png"
          style="height: 20px; width: 20px; margin-right: 20px;"
        />
        <p>
          Iniciar brincadeira!
        </p>
      </button>
      <img
        alt="sacola"
        src="sacolas.png"
      />
    </div>
  </div>
</div>
`;
